// api-client/lib/api_client.js
// Generated by CoffeeScript 1.7.1
(function() {
  var ApiClient, EventEmitter, config_dir, default_config, each, extend, find, fs, request, url, util, _ref;

  fs = require('fs');

  config_dir = process.cwd() + "/config";

  if (!fs.existsSync(config_dir)) {
    fs.mkdirSync(config_dir);
  } else if (!fs.statSync(config_dir).isDirectory()) {
    raise("" + config_dir + " must be a directory, not a regular file");
  }

  url = require('url');

  default_config = require('config');

  util = require('util');

  request = require('request');

  _ref = require('underscore'), extend = _ref.extend, each = _ref.each, find = _ref.find;

  EventEmitter = require('events').EventEmitter;

  ApiClient = (function() {
    ApiClient.create = function(name) {
      var clazz, endpoint_config;
      if (this.config == null) {
        throw new Error("ApiClient not configured");
      }
      endpoint_config = this.config.endpoints[name];
      if (endpoint_config == null) {
        throw new Error("ApiClient endpoint " + name + " not configured");
      }
      clazz = ApiClient;
      if (endpoint_config.type) {
        clazz = this.types[endpoint_config.type];
      }
      return new clazz(endpoint_config);
    };

    ApiClient.default_config = default_config;

    ApiClient.default_request_options = {
      timeout: 2000
    };

    ApiClient.logger = null;

    ApiClient.types = {
      'ApiClient': ApiClient
    };

    ApiClient.load = function(config, dirname) {
      var files;
      if (dirname == null) {
        dirname = __dirname;
      }
      if (this.config) {
        return;
      }
      this.config = config || this.default_config;
      files = fs.readdirSync(dirname);
      return each(files, function(file) {
        var full_path;
        full_path = "" + dirname + "/" + file;
        return require(full_path);
      });
    };

    ApiClient.register = function(label, clazz, clazz_name, endpoint_config) {
      var _base;
      this.types || (this.types = {});
      this.types[clazz_name] = clazz;
      this.config || (this.config = {});
      (_base = this.config).endpoints || (_base.endpoints = {});
      return this.config.endpoints[label] = endpoint_config;
    };

    function ApiClient(options) {
      this.host = options.host;
      this.port = options.port;
      this.options = options.options || {};
      if (this.options.protocol && this.options.protocol === 'https') {
        this.port || (this.port = 443);
      } else {
        this.port || (this.port = 80);
      }
      this.request_options = extend({}, ApiClient.default_request_options, options.request_options);
      each(options.stubs || [], (function(_this) {
        return function(stub) {
          return _this.stub_request(stub);
        };
      })(this));
    }

    ApiClient.prototype.api_path = function() {
      return this.options.base_path || "/";
    };

    ApiClient.prototype.url_config = function(params) {
      if (params == null) {
        params = {};
      }
      return {
        hostname: this.host,
        port: this.port,
        pathname: this.api_path(),
        protocol: this.options.protocol || 'http',
        query: params
      };
    };

    ApiClient.prototype.url = function(params) {
      if (params == null) {
        params = {};
      }
      return url.format(this.url_config(params));
    };

    ApiClient.prototype.stub_request = function(_arg) {
      var body, err, response, uriRegex;
      uriRegex = _arg.uriRegex, err = _arg.err, response = _arg.response, body = _arg.body;
      if (body != null ? body.file : void 0) {
        body = fs.readFileSync(body.file).toString();
      }
      return this.stubs().push([uriRegex, err, response, body]);
    };

    ApiClient.prototype.stubs = function() {
      return this.stubArray || (this.stubArray = []);
    };

    ApiClient.prototype.get = function(params, headers, cb) {
      var e, request_opts, start, stub;
      if (cb == null) {
        cb = void 0;
      }
      request_opts = {
        uri: this.url(params),
        headers: headers,
        method: 'GET'
      };
      if (this.request_options != null) {
        extend(request_opts, this.request_options);
      }
      stub = find(this.stubs(), function(stub) {
        var pattern;
        pattern = stub[0];
        if (!(pattern instanceof RegExp)) {
          pattern = new RegExp(pattern);
        }
        return pattern.test(request_opts.uri);
      });
      if (stub) {
        if (cb) {
          process.nextTick(function() {
            return cb(stub[1], stub[2], stub[3]);
          });
          return null;
        } else {
          e = new EventEmitter;
          process.nextTick(function() {
            if (stub[1] != null) {
              return e.emit('error', stub[1]);
            } else {
              return e.emit('complete', stub[2], stub[3]);
            }
          });
          return e;
        }
      }
      start = new Date;
      if (cb) {
        request_opts.callback = function(err, response, body) {
          var delta, stop;
          stop = new Date;
          delta = stop.getTime() - start.getTime();
          if (ApiClient.logger) {
            ApiClient.logger.info("[APICLIENT] request stop: " + response.statusCode + ": " + request_opts.uri + ", delta: " + delta);
          }
          return cb(err, response, body);
        };
      }
      if (this.options.username && this.options.password) {
        extend(request_opts, {
          user: this.options.username,
          pass: this.options.password
        });
      }
      if (ApiClient.logger) {
        ApiClient.logger.info("[APICLIENT] request start: " + request_opts.uri);
      }
      return this.request(request_opts);
    };

    ApiClient.prototype.request = request;

    return ApiClient;

  })();

  module.exports = ApiClient;

}).call(this);
